File: src/engine/fse/BufferReference.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SimpleEngine.fse
{
    public class BufferReference
    {
        Fbo buffer = null;
        readonly string name;

        public BufferReference(string name)
        {
            this.name = name;
        }

        public string Name
        {
            get
            {
                return name;
            }
        }

        internal void setBuffer(Fbo fbo)
        {
            if (buffer == null) buffer = fbo;
            else throw new Exception("Error, buffer already initialized!");
        }

        internal void bindTexture(int location)
        {
            buffer.bindTexture(location);
        }

        internal void updateTexture(Action a)
        {
            buffer.updateTexture(a);
        }

        internal int Width
        {
            get
            {
                return buffer.Width;
            }
        }

        internal int Height
        {
            get
            {
                return buffer.Height;
            }
        }
    }
}

File: src/engine/fse/bufferreference.cc
#include "engine/fse/bufferreference.h"

#include <stdexcept>

namespace simple::fse
{
    BufferReference::BufferReference(const std::string& name)
        : name(name)
    {
    }

    void BufferReference::SetBuffer(std::shared_ptr<Fbo> fbo)
    {
        if (buffer == nullptr)
        {
            buffer = fbo;
        }
        else
        {
            throw std::runtime_error("Error, buffer already initialized!");
        }
    }

    void BufferReference::BindTexture(int location)
    {
        buffer->BindTexture(location);
    }

    void BufferReference::UpdateTexture(std::function<void()> a)
    {
        buffer->UpdateTexture(std::move(a));
    }

    int BufferReference::GetWidth() const
    {
        return buffer->Width();
    }

    int BufferReference::GetHeight() const
    {
        return buffer->Height();
    }
}
